server/src/main/java/io/littlehorse/common/model/metadatacommand/subcommand/PutStructDefRequestModel.java

```diff
@@ -0,0 +1,134 @@
+    }
+
+    @Override
+    public void initFrom(Message p, ExecutionContext context) throws LHSerdeException {
+        PutStructDefRequest proto = (PutStructDefRequest) p;
+
+        name = proto.getName();
+        structDef = LHSerializable.fromProto(proto.getStructDef(), InlineStructDefModel.class, context);
+        allowedUpdateType = proto.getAllowedUpdates();
+    }
+
+    @Override
+    public boolean hasResponse() {
+        return true;
+    }
+
+    @Override
+    public Message process(MetadataCommandExecution context) {
+        MetadataManager metadataManager = context.metadataManager();
+        this.validate();
+
+        StructDefModel spec = new StructDefModel(context);
+        spec.setId(new StructDefIdModel(name, 0));
+        spec.setStructDef(structDef);
+        spec.setCreatedAt(new Date());
+
+        StructDefModel oldVersion = context.service().getStructDef(name, null);
+
+        if (oldVersion != null) {
+            if (StructDefUtil.equals(spec, oldVersion)) {
+                return oldVersion.toProto().build();
+            }
+
+            verifyUpdateType(allowedUpdateType, spec, oldVersion);
+            spec.getObjectId().setVersion(oldVersion.getObjectId().getVersion() + 1);
+        }
+
+        metadataManager.put(spec);
+        return spec.toProto().build();
+    }
+
+    private void verifyUpdateType(
+            AllowedStructDefUpdateType allowedUpdateType, StructDefModel newSpec, StructDefModel oldSpec) {
+        switch (allowedUpdateType) {
+            case FULLY_COMPATIBLE_SCHEMA_UPDATES:
+                validateFullyCompatibleSchema(newSpec, oldSpec);
+                break;
+            case NO_SCHEMA_UPDATES:
+                throw new LHApiException(
+                        Status.INVALID_ARGUMENT,
+                        "StructDef [%s] already exists and cannot be updated with the selected compatiblity type NO_SCHEMA_UPDATES."
+                                .formatted(name));
+            case UNRECOGNIZED:
+            default:
+                break;
+        }
+    }
+
+    private void validateFullyCompatibleSchema(StructDefModel newSpec, StructDefModel oldSpec) {
+        List<Entry<String, StructFieldDefModel>> changedFields = StructDefUtil.getBreakingChanges(newSpec, oldSpec);
+
+        if (changedFields.isEmpty()) return;
+
+        StringBuilder errorMessage = new StringBuilder("Incompatible schema evolution on field(s): ");
+
+        String fieldNames = changedFields.stream()
+                .map(Entry::getKey)
+                .map((keyStr) -> "[%s]".formatted(keyStr))
+                .collect(Collectors.joining(", "));
+
+        errorMessage.append(fieldNames);
+
+        throw new LHApiException(Status.INVALID_ARGUMENT, errorMessage.toString());
+    }
+
+    @Override
+    public Class<PutStructDefRequest> getProtoBaseClass() {
+        return PutStructDefRequest.class;
+    }
+
+    public void validate() {
+        if (!LHUtil.isValidLHName(name)) {
```

Comment:
Small nitpick. Is it necessary to put this in a helper method? It kind of threw me off since I thought `validate()` was going to do the compatibility checking. I think (since this method is so short) we can just put the lines directly in `process()`